cylc/flow/network/graphql.py

```diff
@@ -128,200 +149,88 @@ def strip_null(data):
-                    components['parent_type'],
-                    arg_name,
-                    arg_value,
-            ):
-                return True
-        return False
-
-    def args_selection_search(
-            self, selection_set, variables, parent_type, arg_name, arg_value):
-        """Recursively search through feild/fragment selection set fields."""
-        for field in selection_set.selections:
-            if isinstance(field, ast.FragmentSpread):
-                if field.name.value in self.visited_fragments:
-                    continue
-                frag_def = self.fragment_defs[field.name.value]
-                frag_type = type_from_ast(self.schema, frag_def.type_condition)
-                if self.args_selection_search(
-                        frag_def.selection_set, variables,
-                        frag_type, arg_name, arg_value):
-                    return True
-                self.visited_fragments.add(frag_def.name)
-                continue
-            field_def = get_field_def(
-                self.schema, parent_type, field.name.value)
-            if field_def is None:
-                continue
-            arg_vals = get_argument_values(
-                field_def.args, field.arguments, variables)
-            if arg_vals.get(arg_name) == arg_value:
-                return True
-            if field.selection_set is None:
-                continue
-            if self.args_selection_search(
-                    field.selection_set, variables,
-                    get_named_type(field_def.type), arg_name, arg_value):
-                return True
-        return False
-
-
-def execute_and_validate_and_strip(
-    schema: 'GraphQLSchema',
-    document_ast: 'Document',
-    *args: Any,
-    **kwargs: Any
-) -> Union['ExecutionResult', Observable]:
-    """Wrapper around graphql ``execute_and_validate()`` that adds
-    null stripping."""
-    result = execute_and_validate(schema, document_ast, *args, **kwargs)
-    # Search request document to determine if 'stripNull: true' is set
-    # as and argument. It can not be done in the middleware, as they
-    # can be Promises/futures (so may not been resolved at this point).
-    variable_values = kwargs['variable_values'] or {}
-    doc_args = AstDocArguments(schema, document_ast, variable_values)
-    if doc_args.has_arg_val(STRIP_ARG, True):
-        if kwargs.get('return_promise', False) and hasattr(result, 'then'):
-            return result.then(null_stripper)  # type: ignore[union-attr]
-        return null_stripper(result)
-    return result
-
-
-class CylcGraphQLBackend(GraphQLBackend):
-    """Return a GraphQL document using the default
-    graphql executor with optional null-stripping of result.
-
-    The null value stripping of result is triggered by the presence
-    of argument & value "stripNull: true" in any field.
+    if is_awaitable(exe_result):
+        return async_next(strip_null, exe_result)
```

Comment:
Is it possible for `exe_result` to be awaitable?

I tried jamming a breakpoint in here, but I've not yet managed to get the code to hit it.


I tried jamming a breakpoint on this line

```diff
diff --git a/cylc/flow/network/graphql.py b/cylc/flow/network/graphql.py
index d2a311989..fe3dab70b 100644
--- a/cylc/flow/network/graphql.py
+++ b/cylc/flow/network/graphql.py
@@ -152,6 +152,7 @@ def strip_null(data):
 def null_stripper(exe_result):
     """Strip nulls in accordance with type of execution result."""
     if is_awaitable(exe_result):
+        breakpoint()
         return async_next(strip_null, exe_result)
     return strip_null(exe_result)
 
```

Reply:
Perhaps not with the current field executor...

Reply:
Trying to work out whether it makes sense to keep these code branches, I don't think they've been tested?